### PR TITLE
Halite helpers

### PR DESCRIPTION
**Pull request by harrisse**
REVIEWERS' NOTE: This PR is kind of a monster, apologies in advance. I've annotated each file with some commentary about what's worth reading and what's worth skipping, large parts of the PR are boring refactors on which I welcome comments but are not likely the best use of your time.

Feel free to skip any parts you don't feel are relevant to your interests. We are on a relatively tight timeline to get this out before Halite so if there are any larger pieces of feedback I'll likely save those for a future iteration. If you don't have time to read through this, then no worries and feel free to disregard the PR.

This PR provides wrapper classes for the halite environment and state as well as a convenient hook for the interpreter. This also reimplements the bulk of the original halite interpreter in the new board class, straightens out some logic inconsistencies in corner cases of the old implementation, and hopefully makes it a bit easier to parse what's happening in the code.

I tested the interpreter reimplementation by writing unit tests and by running actual episodes using the agents submitted to the shared repo. I can't validate that behavior is exactly the same but I can at least confirm that everyone's agents played complete and reasonably strategic-looking games that seemed similar or identical to the current implementation.

### CODE REVIEW

File: kaggle_environments/envs/halite/halite.py [hunk 1]
```diff
@@ -36,319 +38,133 @@ def get_to_pos(size, pos, direction):
...
-            
-    # Spawn ships (30% probability when possible, or 100% if no ships).
-    for uid, pos in shipyards.items():
-        if board.ships[pos] is None and player_halite >= config.spawnCost and (randint(0, 2) == 2 or len(ships_items) == 0):
-            player_halite -= config.spawnCost
-            board.spawn(uid)
-
-    return board.action
+@board_agent
+def random_agent(board):
```
Comment (harrisse) [inline kaggle_environments/envs/halite/halite.py:42]: Random weights have been tweaked for readability as random is meant to be an example for users not a strong bot.

File: kaggle_environments/envs/halite/halite.py [hunk 2]
```diff
@@ -36,319 +38,133 @@ def get_to_pos(size, pos, direction):
...
+    for shipyard in shipyards:
+        # 20% chance to spawn
+        shipyard.next_action = choice([ShipyardAction.SPAWN, None, None, None, None])
 
 
 agents = {"random": random_agent}
 
 
-def interpreter(state, env):
+def populate_board(state, env):
```
Comment (harrisse) [inline kaggle_environments/envs/halite/halite.py:57]: Everything in the populate_board method is the same as it was before, just refactored into its own method. Reviewers can skip to the end of this method.

File: kaggle_environments/envs/halite/halite.py [hunk 3]
```diff
@@ -36,319 +38,133 @@ def get_to_pos(size, pos, direction):
...
-
-        # Initialize the players.
-        obs.players = []
-        for i in range(num_agents):
-            ships = {create_uid(): [starting_positions[i], 0]}
-            obs.players.append([state[0].reward, {}, ships])
-
-        return state
-
-    # Apply actions to create an updated observation.
```
Comment (harrisse) [inline kaggle_environments/envs/halite/halite.py:248]: Reviewers should continue reading here.

All observation update logic is now handled in the Board class, this includes applying actions, resolving collisions, gathering halite, depositing halite, and regenerating halite. Logic to eliminate players or otherwise control the meta-state of the game remains in this method.

File: kaggle_environments/envs/halite/halite.py [hunk 4]
```diff
@@ -36,319 +38,133 @@ def get_to_pos(size, pos, direction):
...
-                    obs.players[player_index][2][uid][1] += smallest_ships[i+1][2]
-        # Detect Shipyard Collisions.
-        if shipyard > -1:
-            enemy_occupied = any(x != shipyard for x in ships.values())
-            if enemy_occupied:
-                del obs.players[shipyard][1][shipyard_uid]
-                for uid, index in list(ships.items()):
-                    del ships[uid]
-                    del obs.players[index][2][uid]
+def interpreter(state, env):
```
Comment (harrisse) [inline kaggle_environments/envs/halite/halite.py:145]: Reviewers should continue reading here.

All observation update logic is now handled in the Board class, this includes applying actions, resolving collisions, gathering halite, depositing halite, and regenerating halite. Logic to eliminate players or otherwise control the meta-state of the game remains in this method.

File: kaggle_environments/envs/halite/helpers.py [hunk 5]
```diff
@@ -0,0 +1,781 @@
...
+    """
+    y, x = divmod(index, size)
+    return x, (size - y - 1)
+
+
+TElement = TypeVar('TElement')
+THash = TypeVar('TComparable')
+
+
+def group_by(elements: Iterable[TElement], selector: Callable[[TElement], THash]) -> Dict[THash, List[TElement]]:
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:47]: Needed to bring a friend in from Linq 😄

File: kaggle_environments/envs/halite/helpers.py [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+            results[key] = []
+        results[key].append(element)
+    return results
+
+
+TKey = TypeVar('TKey')
+TValue = TypeVar('TValue')
+
+
+class ReadOnlyDict(Generic[TKey, TValue]):
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:61]: Everything from here to the board class is mainly just data binding models. Reviewers can mostly skim these, but there are some small bits of logic like `ShipAction.to_point` and `Player.next_actions`

File: kaggle_environments/envs/halite/helpers.py [hunk 7]
```diff
@@ -0,0 +1,781 @@
...
+    def observation(self):
+        """Converts a player back to the normalized observation subset that constructed it."""
+        shipyards = {shipyard.id: shipyard.observation for shipyard in self.shipyards}
+        ships = {ship.id: ship.observation for ship in self.ships}
+        return [self.halite, shipyards, ships]
+# endregion
+
+
+class Board:
+    def __init__(
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:460]: The constructor is basically just used to deserialize the raw observation and configuration received from the halite interpreter into the richer Board model that we'll make use of in the next() method down below.

File: kaggle_environments/envs/halite/helpers.py [hunk 8]
```diff
@@ -0,0 +1,781 @@
...
+    def opponents(self) -> List[Player]:
+        """
+        Returns all players that aren't the current player.
+        You can get all opponent ships with [ship for ship in player.ships for player in board.opponents]
+        """
+        return [player for player in self.players.values() if not player.is_current_player]
+
+    @property
+    def observation(self) -> Dict[str, Any]:
+        """Converts a Board back to the normalized observation that constructed it."""
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:574]: This is the observation serializer to the constructor's observation deserializer

File: kaggle_environments/envs/halite/helpers.py [hunk 9]
```diff
@@ -0,0 +1,781 @@
...
+        return {
+            "halite": halite,
+            "players": players,
+            "player": self.current_player_id,
+            "step": self.step
+        }
+
+    def __deepcopy__(self, _) -> 'Board':
+        actions = [player.next_actions for player in self.players.values()]
+        return Board(self.observation, self.configuration, actions)
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:588]: Because the transformation to an observation copies all of the data into new data structures we can use it for deepcopy as well to isolate a copied instance from its parent.

File: kaggle_environments/envs/halite/helpers.py [hunk 10]
```diff
@@ -0,0 +1,781 @@
...
+    def __str__(self) -> str:
+        """
+        The board is printed in a grid with the following rules:
+        Capital letters are shipyards
+        Lower case letters are ships
+        Digits are cell halite and scale from 0-9 directly proportional to a value between 0 and self.configuration.max_cell_halite
+        Player 1 is letter a/A
+        Player 2 is letter b/B
+        etc.
+        """
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:606]: I'm curious if people think this belongs in `__str__` or another method like `print_board` or something like that?

File: kaggle_environments/envs/halite/helpers.py [hunk 11]
```diff
@@ -0,0 +1,781 @@
...
+                result += str(normalized_halite)
+                result += (
+                    chr(ord('A') + cell.shipyard.player_id)
+                    if cell.shipyard is not None
+                    else ' '
+                )
+            result += '|\n'
+        return result
+
+    def _add_ship(self: 'Board', ship: Ship) -> None:
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:628]: Undocumented as these are meant to be internal only

File: kaggle_environments/envs/halite/helpers.py [hunk 12]
```diff
@@ -0,0 +1,781 @@
...
+            shipyard.cell._shipyard_id = None
+        del self._shipyards[shipyard.id]
+
+    def next(self) -> 'Board':
+        """
+        Returns a new board with the current board's next actions applied.
+        The current board is unmodified.
+        This can form a halite interpreter, e.g.
+            next_observation = Board(current_observation, configuration, actions).next.observation
+        """
```
Comment (harrisse) [inline kaggle_environments/envs/halite/helpers.py:657]: This is the replacement for the observation update logic in the halite interpreter.

File: kaggle_environments/main.py [hunk 13]
```diff
@@ -63,6 +63,9 @@
 parser.add_argument(
     "--host", type=str, help="http-server Host (default=127.0.0.1)."
 )
+parser.add_argument(
+    "--out", type=str, help="Output file to write the results of the episode."
```
Comment (harrisse) [inline kaggle_environments/main.py:67]: This allows you to write the results to an html file while still getting print statements in std out.

File: kaggle_environments/envs/halite/test_halite.py [hunk 14]
```diff
@@ -25,9 +26,121 @@ def test_halite_completes():
...
+    board = create_board(starting_halite=1000, agent_count=1)
+    cell = first(board.cells.values())
+    next_board = board.next()
+    next_cell = next_board[cell.position]
+    expected_regen = round(cell.halite * board.configuration.regen_rate, 3)
+    # We compare to a floating point value here to handle float rounding errors
+    assert next_cell.halite - cell.halite - expected_regen < .000001
+
+
+def test_no_move_on_halite_gathers_halite():
```
Comment (harrisse) [inline kaggle_environments/envs/halite/test_halite.py:136]: There are lots of other scenarios we could test in here (like multiship collisions on shipyards, spawning collisions, etc) but I tried to cover most core rules with this pass.

File: kaggle_environments/utils.py [hunk 15]
```diff
@@ -113,12 +114,16 @@ def get_exec(raw, fallback=None):
...
+        output = buffer.getvalue()
+        if output:
+            print(output)
         return env
     except Exception as e:
         sys.stdout = orig_out
-        print(buffer.getvalue())
-        if fallback != None:
+        output = buffer.getvalue()
+        if output:
```
Comment (harrisse) [inline kaggle_environments/utils.py:124]: This change gets rid of most of those annoying blank lines in the output

File: kaggle_environments/utils.py [hunk 16]
```diff
@@ -15,6 +15,7 @@
 import json
```
Comment (harrisse) [inline kaggle_environments/utils.py:15]: This file is all style and formatting updates, reviewers can likely skip this file.

File: kaggle_environments/envs/halite/helpers.py [hunk 17]
```diff
@@ -0,0 +1,780 @@
...
+                    ship.cell._ship_id = None
+                    ship._position = wrap_point(translate_point(ship.position, ship.next_action.to_point()), configuration.size)
+                    # We don't set the new cell's ship_id here as it would be overwritten by another ship in the case of collision.
+                    # Later we'll iterate through all ships and re-set the cell._ship_id as appropriate.
+                # Clear the ship's action so it doesn't repeat the same action automatically
+                ship.next_action = None
+
+            player._halite += leftover_convert_halite
+
+        def resolve_collision(ships: List[Ship]) -> Tuple[Optional[Ship], List[Ship]]:
```
Comment (moserware) [inline kaggle_environments/envs/halite/helpers.py:705]: These methods are specific to the game's rules and could be `staticmethod`? I wonder if it's worth factoring out to a separate `class` for clarity?

File: kaggle_environments/envs/halite/helpers.py [hunk 18]
```diff
@@ -0,0 +1,780 @@
...
+            elif ship.player_id == shipyard.player_id:
+                # Ship depositing halite
+                shipyard.player._halite += ship.halite
+                ship._halite = 0
+            else:
+                # Ship to shipyard collision
+                board._delete_shipyard(shipyard)
+                board._delete_ship(ship)
+
+        # Collect halite from cells into ships
```
Comment (moserware) [inline kaggle_environments/envs/halite/helpers.py:749]: ❓ Just to confirm, if a smaller ship collides with a bigger ship, it should also collect in that same turn?

File: kaggle_environments/envs/halite/helpers.py [hunk 19]
```diff
@@ -0,0 +1,780 @@
...
+        for ship in board.ships.values():
+            cell = ship.cell
+            delta_halite = int(cell.halite * configuration.collect_rate)
+            if cell.shipyard_id is None and delta_halite > 0:
+                ship._halite += delta_halite
+                cell._halite -= delta_halite
```
Comment (moserware) [inline kaggle_environments/envs/halite/helpers.py:755]: might want another `assert cell._halite > 0` after this? Just to confirm, a cell cannot run out of halite?

File: kaggle_environments/envs/halite/helpers.py [hunk 20]
```diff
@@ -0,0 +1,780 @@
...
+    Automatically returns the modified board's next actions
+
+    @board_agent
+    def my_agent(board: Board) -> None:
+        ...
+    """
+    def agent_wrapper(obs, config) -> Dict[str, str]:
+        board = Board(obs, config)
+        agent(board)
+        return board.current_player.next_actions
```
Comment (moserware) [inline kaggle_environments/envs/halite/helpers.py:779]: Am assuming the JSON type/domain checking is already verifying these actions are valid.

File: kaggle_environments/utils.py (hunk 15 above)
Comment (moserware) [reply kaggle_environments/utils.py:124]: 👏

File: kaggle_environments/envs/halite/helpers.py (hunk 9 above)
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:587]: `next_actions` is a computed property constructed from the player's ships and shipyards so we don't have to worry about modification

**Comment by harrisse** [comment]
@moserware Re: internal only methods and exploits, no bots can't use them unless users are running untrusted code locally (which we ultimately can't protect against). In production we isolate agents in different containers so there's no risk of them being able to call each others' methods.

File: kaggle_environments/envs/halite/helpers.py [hunk 21]
```diff
@@ -0,0 +1,780 @@
...
+    return x % size, y % size
+
+
+def position_to_index(point: Point, size: int) -> int:
+    """
+    Converts a 2d position in the form (x, y) to an index in the observation.halite list.
+    See index_to_position for the inverse.
+    """
+    x, y = point
+    return (size - y - 1) * size + x
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:31]: This puts 0 in the lower left corner and (size - 1, size - 1) in the upper right which I personally found a lot more intuitive, curious to get others' thoughts on this though.

File: kaggle_environments/envs/halite/helpers.py [hunk 22]
```diff
@@ -0,0 +1,780 @@
...
+    Observation primarily used as a helper to construct the Board from the raw observation.
+    This provides bindings for the observation type described at https://github.com/Kaggle/kaggle-environments/blob/master/kaggle_environments/envs/halite/halite.json
+    """
+    @property
+    def halite(self) -> List[float]:
+        """Serialized list of available halite per cell on the board."""
+        return self["halite"]
+
+    @property
+    def players(self) -> List[List[int]]:
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:100]: Json doesn't have a tuple representation so we represent them as fixed length lists in json which become normal lists in python.

File: kaggle_environments/envs/halite/helpers.py [hunk 23]
```diff
@@ -0,0 +1,780 @@
...
+
+    @property
+    def run_timeout(self) -> float:
+        """Maximum runtime (seconds) of an episode (not necessarily DONE)."""
+        return self["runTimeout"]
+
+    @property
+    def halite(self) -> int:
+        """The starting amount of halite available on the board."""
+        return self["halite"]
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:148]: Yes, great catch, luckily we don't use this class for board generation but I'll make sure to get this updated.

File: kaggle_environments/envs/halite/helpers.py [hunk 24]
```diff
@@ -0,0 +1,780 @@
...
+class Cell:
+    def __init__(self, position: Point, halite: float, shipyard_id: Optional[ShipyardId], ship_id: Optional[ShipId], board: 'Board') -> None:
+        self._position = position
+        self._halite = halite
+        self._shipyard_id = shipyard_id
+        self._ship_id = ship_id
+        self._board = board
+
+    @property
+    def position(self) -> Point:
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:234]: Yes, the terms are position and index consistently across helpers.

[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:234]: (Or at least they should be :))

File: kaggle_environments/envs/halite/helpers.py [hunk 25]
```diff
@@ -0,0 +1,780 @@
...
+            for ship_id in self.ship_ids
+        ]
+
+    @property
+    def is_current_player(self) -> bool:
+        """Returns whether this player is the current player (generally if this returns True, this player is you)."""
+        return self.id is self._board.current_player_id
+
+    @property
+    def next_actions(self) -> Dict[str, str]:
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:435]: I think Dict[str, str] is correct because that's the type expected from an agent response and this field is meant to be returnable from agents.

File: kaggle_environments/envs/halite/helpers.py [hunk 26]
```diff
@@ -0,0 +1,780 @@
...
+        https://github.com/Kaggle/kaggle-environments/blob/master/kaggle_environments/envs/halite/halite.json
+        Board tracks players (by id), ships (by id), shipyards (by id), and cells (by position).
+        Each entity contains both key values (e.g. ship.player_id) as well as entity references (e.g. ship.player).
+        References are deep and chainable e.g.
+            [ship.halite for player in board.players for ship in player.ships]
+            ship.player.shipyards[0].cell.north.east.ship
+        Consumers should not set or modify any attributes except Ship.next_action and Shipyard.next_action
+        """
+        observation = Observation(raw_observation)
+        # Pending actions is effectively a Dict[Union[[ShipId, ShipAction], [ShipyardId, ShipyardAction]]]
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:476]: Should be next, thanks, I'll make sure to ctrl+f pending.

File: kaggle_environments/envs/halite/helpers.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (dster2) [reply kaggle_environments/envs/halite/helpers.py:31]: Not sure I'm fully understanding, but my personal take is that `(0, 0)` as top left makes more sense because that's how 2d arrays work, and in particular the format used in the replay files.  I see how your way is intuitive relative to normal XY plane but IMO it's better to match the programming methodology.  But just a personal take and IIUC a lot / all of this is internal and the higher-level structure here means users aren't necessarily forced into either paradigm?

[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:31]: @dster2 Yeah users would only even notice the coordinate system if they were constructing offsets by hand. If they only used NORTH/SOUTH/EAST/WEST they'd likely never even notice there are coordinates. I'll consider this a bit more, thanks for the feedback all.

File: kaggle_environments/envs/halite/helpers.py [hunk 27]
```diff
@@ -0,0 +1,780 @@
...
+TValue = TypeVar('TValue')
+
+
+class ReadOnlyDict(Generic[TKey, TValue]):
+    """Offers Dict-like semantics while preventing modification of the underlying datastructure."""
+    def __init__(self, data: Union['ReadOnlyDict[TKey, TValue]', Dict[TKey, TValue]]):
+        if isinstance(data, dict):
+            self._data = data
+        else:
+            # If it's not a Dict it must be a ReadOnlyDict based on our type
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:67]: My goal here wasn't to validate that it must be a readonlydict but that if the user passed in the correct type and it's not a dict then it must be a readonlydict.

File: kaggle_environments/envs/halite/helpers.py [hunk 28]
```diff
@@ -0,0 +1,780 @@
...
+        self._step = observation.step
+        self._configuration = Configuration(raw_configuration)
+        self._current_player_id = observation.player
+        self._players: Dict[PlayerId, Player] = {}
+        self._ships: Dict[ShipId, Ship] = {}
+        self._shipyards: Dict[ShipyardId, Shipyard] = {}
+        self._cells: Dict[Point, Cell] = {}
+
+        size = self.configuration.size
+        # Create a cell for every point in a size x size grid
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:490]: Because about half of all cells have halite, sparse would still give us O(n^2), but fair point that it would cut down on a number of objects.

File: kaggle_environments/envs/halite/helpers.py [hunk 29]
```diff
@@ -0,0 +1,780 @@
...
+        You can get all opponent ships with [ship for ship in player.ships for player in board.opponents]
+        """
+        return [player for player in self.players.values() if not player.is_current_player]
+
+    @property
+    def observation(self) -> Dict[str, Any]:
+        """Converts a Board back to the normalized observation that constructed it."""
+        size = self.configuration.size
+        halite = [self[index_to_position(index, size)].halite for index in range(size * size)]
+        players = [player._observation for player in self.players.values()]
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:576]: I confirmed that list and generator ordering are explicitly preserved.

File: kaggle_environments/envs/halite/helpers.py [hunk 30]
```diff
@@ -0,0 +1,780 @@
...
+        for y in range(size):
+            for x in range(size):
+                cell = self[(x, size - y - 1)]
+                result += '|'
+                result += (
+                    chr(ord('a') + cell.ship.player_id)
+                    if cell.ship is not None
+                    else ' '
+                )
+                normalized_halite = int(9.0 * cell.halite / float(self.configuration.max_cell_halite))
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:617]: This is just to normalize the halite value to a digit between 0-9 so we can print it in a single character on the board.

File: kaggle_environments/envs/halite/helpers.py (hunk 20 above)
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:779]: These get filtered out and ignored in the board constructor if they're not valid

File: kaggle_environments/envs/halite/helpers.py (hunk 18 above)
Comment (harrisse) [reply kaggle_environments/envs/halite/helpers.py:750]: I'll double check the rules to make sure they're consistent with this behavior but I believe so
